### PR TITLE
De-dup code: use the initialiseState function

### PR DESCRIPTION
**Pull request by dbkr**
This should behave identically, but the code here appeared to be
identical to the code in initialiseState, so let's use it (it also
has an extra sanity check in there that we only init empty timelines).

Edit: I take back what I said about this being functionally identical: it isn't. It appears that previously we were applying any events in the `state` field of the join room to both the old state and the new state, which is surely wrong. Although these are normally only ever sent in a limited or initial sync (so it wouldn't matter) they are well-defined in other syncs as events that precede the timeline events, so we should process them as we would timeline events, before we process the timeline events. The tests send sync responses with state changes in `state` rather than `timeline`.

### CODE REVIEW

File: src/sync.js [hunk 1]
```diff
@@ -1294,41 +1294,35 @@ SyncApi.prototype._resolveInvites = function(room) {
...
-    // will leak through to current state!
-    const oldStateEvents = utils.map(
-        utils.deepCopy(
-            stateEventList.map(function(mxEvent) {
-                return mxEvent.event;
-            }),
-        ), client.getEventMapper(),
-    );
-    const stateEvents = stateEventList;
+    // We'll only get state events in stateEventList if this is a limited
```
Comment (richvdh) [inline src/sync.js]: hrm... this is an assumption based on the way synapse behaves. Indeed it seems to be in conflict with the fact that you then do something with the state events if the timeline is not empty.

Could you rephrase this comment please.

File: src/sync.js [hunk 2]
```diff
@@ -1294,41 +1294,35 @@ SyncApi.prototype._resolveInvites = function(room) {
...
     room.recalculate(this.client.credentials.userId);
 
-    // execute the timeline events, this will begin to diverge the current state
+    // If the timeline wasn't empty, we process the state events here: they're
+    // defined as updates to the state before the start of the timeline, so this
+    // starts to roll the state forward.
+    if (!timelinewasEmpty) {
+        room.addLiveEvents(stateEventList || []);
+    }
+    // execute the timeline events, this will continue to diverge the current state
```
Comment (richvdh) [inline src/sync.js]: s/,/./

File: src/sync.js [hunk 3]
```diff
@@ -1294,41 +1294,35 @@ SyncApi.prototype._resolveInvites = function(room) {
...
-    // XXX: what if we've already seen (some of) the events in the timeline,
-    // and they modify some of the state set in stateEvents? In that case we'll
-    // end up with the state from stateEvents, instead of the more recent state
-    // from the timeline.
-    room.oldState.setStateEvents(oldStateEvents);
-    room.currentState.setStateEvents(stateEvents);
+
+    // If there are no events in the timeline yet, initialise it with
+    // the given state events
+    const timelinewasEmpty = room.getLiveTimeline().getEvents().length == 0;
```
Comment (richvdh) [inline src/sync.js]: `timelineWasEmpty` ?

File: src/sync.js [hunk 4]
```diff
@@ -1294,41 +1294,35 @@ SyncApi.prototype._resolveInvites = function(room) {
...
+    if (timelinewasEmpty) {
+        room.getLiveTimeline().initialiseState(stateEventList);
+    }
 
     this._resolveInvites(room);
```
Comment (richvdh) [inline src/sync.js]: why does this (and recalculate) happen after adding the state events for an empty timeline, but before it for an existing timeline?

File: src/sync.js (hunk 3 above)
Comment (richvdh) [inline src/sync.js]: can we factor out the myriad calls to `room.getLiveTimeline()` into a local var?

**Comment by richvdh** [comment]
also: any chance of some tests for this stuff?

File: src/sync.js (hunk 1 above)
Comment (dbkr) [reply src/sync.js]: tbh I think deleting it may make more sense: I think this was left in from when the code was relying on this behaviour in synapse and I failed to remove it.

File: src/sync.js (hunk 2 above)
Comment (dbkr) [reply src/sync.js]: done

File: src/sync.js (hunk 3 above)
Comment (dbkr) [reply src/sync.js]: mm, yeah - fixed

File: src/sync.js (hunk 4 above)
Comment (dbkr) [reply src/sync.js]: My reasoning for this was the the state events ought to be treated the same as if they'd arrived in the timeline section. That said, I opted not to chase that particular rabbit down the hole to figure out what actually happens on a room name change, etc.

**Comment by dbkr** [comment]
Now with tests that fails on develop

File: src/sync.js (hunk 4 above)
Comment (richvdh) [reply src/sync.js]: well, if that's true, why isn't it true for when the timeline is empty?

"I didn't go down the rabbithole" is valid but an `XXX` comment saying so would be good for future reference

Comment (dbkr) [reply src/sync.js]: good point, and now looking at it, it looks a bit broken. Have added an XXX to explain how I think it might be broken.
